src/app/ConcreteAttributePath.h

```diff
@@ -38,9 +39,22 @@ struct ConcreteAttributePath
         return mEndpointId == other.mEndpointId && mClusterId == other.mClusterId && mAttributeId == other.mAttributeId;
     }
 
+    enum class ListOperation
+    {
+        NotList,
+        ReplaceAll,
+        UpdateItem,
+        DeleteItem,
+        AppendItem
+    };
+
+    bool IsListOperation() const { return mListOp != ListOperation::NotList; }
+
     EndpointId mEndpointId   = 0;
     ClusterId mClusterId     = 0;
     AttributeId mAttributeId = 0;
+    uint16_t mListIndex      = 0;
+    ListOperation mListOp    = ListOperation::NotList;
```

Comment:
The interaction between these two needs to be clearly documented.  In particular, I asume that mListIndex must be ignored if mListOp has certain values?

But also, it seems odd to me to have the list operation stuck into the attribute path.  Seems like the only state in the path should be "do we have a list index, and if so what is it?" and the actual "what are we doing at this path" and "is this a path to a list at all?" bits should be handled by some other data.

Reply:
> In particular, I asume that mListIndex must be ignored if mListOp has certain values?

I'm all for more documentation, but in this case, isn't that fairly apparent the relationship between the two?

> Seems like the only state in the path should be "do we have a list index, and if so what is it?" and the actual "what are we doing at this path" and "is this a path to a list at all?" bits should be handled by some other data.

The spec clearly defines interpretations for what these different permutations of list-index being present and not null, being present and null, and not being present at all, mean. Why wouldn't we follow suit?

More importantly, the alternative of just presenting it sans this interpretation would look fairly clunky IMHO - it would have to contain some enumeration that effectively describes 'list index present and valid, list index present and null, list index not present', making consumers figure out the interpretation (and potentially, getting it wrong).

It also importantly, doesn't convey list deletions (since that is a null data value combined with a valid list index), which would then require clients to have to manually parse the data to figure this out. Again, a chance for them to get it wrong.

All of this feels like duplicative logic that consumers shouldn't have to have to do...

Reply:
Hmm...but this won't work out well for paths in reads though. Certainly when a path is accompanied by data, I think having a list operation makes sense. When it is not accompanied by data, then, it's less sensible.

Will have to think some more...

Reply:
Recommendation: To create two separable concrete path types: `ConcreteReadAttributePath` and `ConcreteDataAttributePath` (or something like that).

Reply:
> I'm all for more documentation, but in this case, isn't that fairly apparent the relationship between the two?

It's not.  That is, it's not clear which values of mListOp might or might not have a meaningful mListIndex. 

For the path, the big question is what concept a ConcreteAttributePath represents.  If it's the IM spec's "concrete path", which is what it's been so far, then that concept does not include "what am I doing with this path?"; that's state separate from the path itself.

The encoding spec does encode some aspects of the "what am I doing" state in the TLV path, but that's not the object ConcreteAttributePath is representing, right?

So here's the real question: will all users of ConcreteAttributePath want this list info?  If not, we should create a new struct, or different structs for read/write server-side cases, so read code doesn't try to figure out what it's supposed to do with the Delete list op.  These structs would include a path (by inheritance or composition) and the extra info about what we are doing with the path, and we would use them in the places that actually need the extra info.

Reply:
I've left a [comment](https://github.com/CHIP-Specifications/connectedhomeip-spec/pull/4618#discussion_r752476021) in Cam's PR where he's adding a ListIndex to AttributePathIB, but falling short of actually covering the use cases of conveying update/delete/append.

> or different structs for read/write server-side cases

Yes, that was my recommendation above.

Reply:
Ah, I did not see those comments before I posted mine.  ;)

So it seems like we have three things:

1. ConcreteAttributePath -- identifies an attribute.
2. ConcreteReadAttributePath -- identifies an attribute or a specific index in a list attribute.
3. ConcreteAttributeDataPath -- identifies a data mutation of some sort.

2 and 3 can inherit from or compose 1.  Seem reasonable?

Reply:
I was thinking the same!